File: Java/src/gestorAplicacion/gestorMusica/MeGusta.java
package gestorAplicacion.gestorMusica;

import java.io.Serializable;
import java.util.ArrayList;
import gestorAplicacion.gestorPersonas.Usuario;

public class MeGusta extends Lista implements Serializable {
	
	private static final long serialVersionUID = 1L;
	private static ArrayList<MeGusta> favoritosExistentes;
	private ArrayList<Cancion> favoritos = new ArrayList<Cancion>();
		
	static {
		setFavoritosExistentes(new ArrayList<MeGusta>());
	}
	
	public MeGusta(Usuario usuario){
		super("Tus Me Gusta", usuario, "Tus canciones favoritas");
		favoritosExistentes.add(this);
	}
	
	@Override
	public String agregarCancion(Cancion cancion) {
		favoritos.add(cancion);
		return "Cancion agregada con exito a tus Me Gusta";
	}
	
	@Override
	public String eliminarCancion(Cancion cancion) {
		favoritos.remove(cancion);
		return "Cancion eliminada con exito de tus Me Gusta";
	}

	public ArrayList<Cancion> getFavoritos() {
		return favoritos;
	}
	
	public void setFavoritos(ArrayList<Cancion> favoritos) {
		this.favoritos = favoritos;
	}

	public static ArrayList<MeGusta> getFavoritosExistentes() {
		return favoritosExistentes;
	}

	public static void setFavoritosExistentes(ArrayList<MeGusta> favoritosExistentes) {
		MeGusta.favoritosExistentes = favoritosExistentes;
	}

	@Override
	public int totalPorGenero(Genero genero) {	
		int total = 0;
		for(Cancion cancion: getFavoritos()) {
			if(cancion.getGenero().equals(genero)) {
				total += 1;
			}
		}
		return total;
	}
	
	@Override
	public String toString() {
		String des = "";
		ArrayList<Cancion> lista = usuario.getFavoritos().getFavoritos();
		if (lista.size() > 0) {
			for (Cancion cancion : lista) {
				des += cancion.descripcion() + "\n";
			}
			return "Tus Canciones Favoritas:" + "\n" + "\n" + "Canciones: " + "\n" + des;
		} else {
			return "No tienes canciones favoritas";
		}
	}
}

File: Python/src/gestorAplicacion/gestorMusica/meGusta.py
from gestorAplicacion.gestorMusica.lista import Lista

class meGusta(Lista):
    
    __favoritosExistentes = []
    
    #falta inicializador estatico
    
    def __init__(self, usuario):
        super().__init__("Tus Me Gusta", usuario, "Tus canciones favoritos")
        self.__favoritos = []
        meGusta.__favoritosExistentes.append(self)
        
    def agregarCancion(self, cancion):
        if cancion in self.__favoritos:
            return f"La canción {cancion.getNombre()} ya está dentro de tus favoritos"
        else:
            self.__favoritos.append(cancion)
            return "Se ha agregado la canción " + cancion.getNombre() + " a tus favoritos con éxito"
    
    def eliminarCancion(self, cancion):
        self.__favoritos.remove(cancion)
        return "Canción eliminada con éxito de tus Me Gusta"
    
    def getFavoritos(self):
        return self.__favoritos
    
    def setFavoritos(self, favoritos):
        self.__favoritos = favoritos
    
    @classmethod
    def getFavoritosExistentes(cls):
        return cls.__favoritosExistentes
    
    @classmethod
    def setFavoritosExistentes(cls, favoritosExistentes):
        cls.__favoritosExistentes = favoritosExistentes
        
    def totalPorGenero(self, genero):
        total = 0
        for cancion in meGusta.getFavoritos(self):
            if(cancion.getGenero() == genero):
                total += 1
        return total
    
    def __str__(self):
        des = ""
        if(len(self.__favoritos) > 0):
            for cancion in self.__favoritos:
                des += cancion.descripcion() + "\n\n"
            return "Tus Canciones Favoritas:" + "\n" + "\n" + "Canciones: " + "\n" + des
        else:
            return "No tienes canciones favoritas"